src/python/dxpy/app_builder.py

```diff
@@ -212,58 +161,199 @@ def upload_resources(src_dir, project=None, folder='/', ensure_upload=False):
+                        # else raise an error
+                        if force_symlinks or is_local:
+                            files.append(subdir_name)
+                        else:
+                            raise AppBuilderException("Cannot include symlinks to directories outside of the resource directory.  %r points to directory %r" % (dir_path, dir_realpath))
+                        
+
+                # Canonicalize the order of files so that we compute the
+                # checksum in a consistent order
+                for filename in sorted(files):
+                    deref_link = False
+                
+                    relative_filename = os.path.join(relative_dirname, filename)
+                    true_filename = os.path.join(dirname, filename)
+
+                    file_stat = os.lstat(true_filename)
+                    # check for a link here, please!
+                    if os.path.islink(true_filename):
+
+                        # Get the pointed-to path
+                        link_target = os.readlink(true_filename)
+                        is_local=(not os.path.isabs(link_target))
+                        
+                        linkpath_filename = os.path.abspath(os.path.join(dirname, link_target))
+
+                        # make sure that the path NEVER extends outside the resources directory!
+                        d,l = os.path.split(link_target)
+                        link_parts = []
+                        while l:
+                            link_parts.append(l)
+                            d,l = os.path.split(d)
+                        curr_path = dirname
+                        for p in reversed(link_parts):
+                            curr_path = os.path.abspath(os.path.join(curr_path, p))
+                            is_local = (is_local and curr_path.startswith(resources_dir))
+
+                        if not (force_symlinks or is_local):
+                            # if we are pointing outside of the directory, then:
+                            # try to get the true stat of the file and make sure
+                            # to dereference the link!
+                            try:
+                                file_stat = os.stat(os.path.join(dirname, link_target))
+                                deref_link = True
+                            except OSError:
+                                # uh-oh! looks like we have a broken link!
+                                # since this is guaranteed to cause problems (and
+                                # we know we're not forcing symlinks here), we
+                                # should throw an error
+                                raise AppBuilderException("Broken symlink: Link '%r' points to '%r', which does not exist" % (true_filename, linkpath_filename) )
+                    
+                    
+                    fields = [relative_filename, str(file_stat.st_mode), str(int(file_stat.st_mtime * 1000))]
```

Comment:
When a link is being dereferenced, should we perhaps take the maximum of the mtimes on the link and on its referent? Otherwise, it seems like you could modify the referent and the code here wouldn't realize anything had changed (and thereby allow use of a stale cached tarball).